### ISSUE TITLE
Problems when try to compile source of flare engine

### ISSUE THREAD

**Issue by SarcrofS** [clintbellanger/flare-engine#1491]
Hello, sorry for asking some "idiot" questions, but I am not a programmer, so I am learning how to compile the source files to use the most new engine of Flare.

I instaled the MinGW with pack of G++ to compile, I configure path in windows to recognize the compiler and open Command Prompt, I execute the code above, learned in [INSTALL.engine](https://github.com/clintbellanger/flare-engine/blob/master/INSTALL.engine.md#one_call_build).

`g++ -I C:\MinGW\include\SDL src\*.cpp src\*.c -o flare.exe -lmingw32 -lSDLmain -lSDL2 -lSDL2_image -lSDL2_mixer -lSDL2_ttf`

The prompt says: 
![prompt](https://cloud.githubusercontent.com/assets/28415584/26084448/80033cf2-39b2-11e7-87fa-2ee29f7907b6.png)
g++ error: src/*.c: Invalid argument

I know, the * refers to all files in the path 'src' using the extension '.c', but I am having this error.

Anyone can help me to make this works? Thanks.

**Comment by dorkster** [comment]
Does the Windows cmd.exe even allow filename globbing/wildcards? A quick Google search leads me to believe no.

You might want to try using [cmder](http://cmder.net/) for a more sane console interface.

**Comment by SarcrofS** [comment]
@dorkster I downloaded the complete version of cmder, make the  "Open Cmder Here" in context menu, restarted the computer, opened the cmder in the paste where the src of flare engine are, make the same command, and the result was the same.
![cmder_result](https://cloud.githubusercontent.com/assets/28415584/26085210/e609991e-39b8-11e7-9315-1dcb5e93c4d2.png)

Please, help me with this, what I need do?

**Comment by dorkster** [comment]
Since cmder uses Bash, you probably need to use forward slashes instead of back slashes.

**Comment by SarcrofS** [comment]
Used forward slashes, still the same error 😢

**Comment by dorkster** [comment]
Hmm, I not sure what's wrong here. I've never used mingw straight from the command line before. To make Windows builds, I use CMake to generate a Makefile and then use mingw's make. Unfortunately, it looks like you had some issues with CMake from your forum post. Maybe a [screenshot of my config](https://lh6.googleusercontent.com/KH3igz-VfCNPYarsumb8m2Si3nGixkOyrRy1tx_OKuT2r1Sa_GP1cNcbJjdj4QxtNvGdWQ6dvZXDJjU=w1366-h645-rw) will help. Keep in mind that I manually copied the SDL headers and libs to my installation of mingw (which CodeBlocks provided).

In the meantime, here's a compiled version of the latest flare-engine master for Windows that you can try: https://drive.google.com/open?id=0B1BVtHygCIkWSk1fQ0VENW1heDg

**Comment by SarcrofS** [comment]
Your screenshot give a error 403.

I am using the engine what you post in OGA forums (https://docs.google.com/uc?id=0B1BVtHygCIkWbjlEdk5wMWhET1U&export=download).

But I want to learn to compile to always maintain the engine updated, since you and others devs always make some changes and implement amazing things, and learn is always good, I will try with Cmaker and make, maybe now this works 😞 

Thanks for everything dorkster, this helps will be posted in credits in my game.

**Comment by dorkster** [comment]
Let me just embed my screenshot instead of linking:
![cmake](https://cloud.githubusercontent.com/assets/913982/26086233/84ba8e9e-39b7-11e7-956d-543a360169f3.png)

Anyway, a "Building for Windows" tutorial should definitely be on the agenda. Even for my own education, since I'd like to remove CodeBlocks from my build environment and just have mingw/cmake.

**Comment by SarcrofS** [comment]
Thanks dorkster, I will try using Cmake and Make.

Yep, a guide for building flare engine will be nice, godly for people what never make this (like me).

Thanks for everything again!

**Comment by SarcrofS** [comment]
OMG now I undestand why src/*.c was a error, because don't have ANY file with .c extension.
Now I downloaded all SDL source, and try to compile. Missing some SDL files, downloaded and now have others errors.

New errors:
![cmder2](https://cloud.githubusercontent.com/assets/28415584/26088054/10061d98-39cb-11e7-9e7c-64e2f67e6ab9.png)

How give permission? I put cmder to execute with administrator (in the true .exe, not in the shortcut).

How fix this? Thanks.

**Comment by dorkster** [comment]
Did you try with `src/*.cpp` instead of `src/*.*`? Otherwise, you're explicitly telling g++ to compile the header files, which you should need to do as they get embedded in the cpp files via the `#include` pre-processor command.

**Comment by SarcrofS** [comment]
> src/PlatformAndroid.cpp:33:17: fatal error: jni.h: No such file or directory
> compilation terminated.
> src/PlatformGCW0.cpp: In function 'void PlatformSetPaths()':
> src/PlatformGCW0.cpp:51:3: error: 'PATH_CONF' was not declared in this scope
>    PATH_CONF = (std::string)getenv("XDG_CONFIG_HOME") + "/flare/";
>    ^
> src/PlatformGCW0.cpp:55:3: error: 'PATH_CONF' was not declared in this scope
>    PATH_CONF = (std::string)getenv("HOME") + "/.config/";
>    ^
> src/PlatformGCW0.cpp:61:3: error: 'PATH_CONF' was not declared in this scope
>    PATH_CONF = "./config/";
>    ^
> src/PlatformGCW0.cpp:64:12: error: 'PATH_CONF' was not declared in this scope
>   createDir(PATH_CONF);

IDK where download this jni.h file, this left make the compilation over :/

**Comment by dorkster** [comment]
Ah, I forgot about the Platform files. Those get embedded in in main.cpp, so you're not supposed to pass them to the compiler. You need to compile everything in FLARE_SOURCES as listed [here](https://github.com/clintbellanger/flare-engine/blob/master/CMakeLists.txt#L95). Without CMake, the easiest way to do this would be to copy those file names to a text file and then substitute `$(cat TEXT_FILE.txt)` where `src/*.cpp` is.

**Comment by SarcrofS** [comment]
Copy like "src/BehaviorAlly.cpp" or "/src/BehaviorAlly.cpp" or "./src/BehaviorAlly.cpp"?

**Comment by dorkster** [comment]
1st and 3rd are good if you're running g++ from the flare-engine directory.

**Comment by SarcrofS** [comment]
```
g++: error: $(cat: No such file or directory
g++: error: flaresources.txt)-I: No such file or directory
```

**Comment by dorkster** [comment]
Oh, apparently cmder doesn't do Bash command substitution. Unfortunately, I think you'll just have to copy and paste the filenames straight into the command line, replacing newlines with spaces.

**Comment by SarcrofS** [comment]
I will try to only left in the src paste the files what are mentioned in the link you provide to me.

**Comment by SarcrofS** [comment]
A problem, if left PlatformWin32.cpp give a error:
```
src/PlatformWin32.cpp: In function 'void PlatformSetPaths()':                         
src/PlatformWin32.cpp:39:3: error: 'PATH_CONF' was not declared in this scope         
   PATH_CONF = PATH_USER = (std::string)getenv("APPDATA") + "\\flare";                
   ^                                                                                  
src/PlatformWin32.cpp:39:15: error: 'PATH_USER' was not declared in this scope        
   PATH_CONF = PATH_USER = (std::string)getenv("APPDATA") + "\\flare";                
               ^                                                                      
src/PlatformWin32.cpp:49:3: error: 'PATH_CONF' was not declared in this scope         
   PATH_CONF = "config";                                                              
   ^                                                                                  
src/PlatformWin32.cpp:50:3: error: 'PATH_USER' was not declared in this scope         
   PATH_USER = "userdata";                                                            
   ^                                                                                  
src/PlatformWin32.cpp:55:12: error: 'PATH_USER' was not declared in this scope        
  createDir(PATH_USER + "\\mods");                                                    
            ^                                                                         
src/PlatformWin32.cpp:58:2: error: 'PATH_DATA' was not declared in this scope         
  PATH_DATA = "";                                                                     
  ^                                                                                   
src/PlatformWin32.cpp:59:16: error: 'CUSTOM_PATH_DATA' was not declared in this scope 
  if (dirExists(CUSTOM_PATH_DATA)) PATH_DATA = CUSTOM_PATH_DATA;                      
                ^                                                                     
src/PlatformWin32.cpp:65:2: error: 'PATH_CONF' was not declared in this scope         
  PATH_CONF = PATH_CONF + "/";                                                        
  ^             
```                                               
If try without this (this .cpp isn't listed in your link) the compilation ends because are lefting.

**Comment by SarcrofS** [comment]
I will go sleep @dorkster thanks for answering me A LOT and of all attention.

Btw, thanks for compile the latest version of engine. You are a amazing guy, keep being this! 👍

**Comment by dorkster** [comment]
Yeah, I must have definitely broken the non-CMake build process somewhere along the line. I'll have to sort that out.

**Comment by SarcrofS** [comment]
Sorry for consuming your time a lot, thanks again, you have my respect.

**Comment by dorkster** [comment]
Non-CMake builds should be working now with https://github.com/clintbellanger/flare-engine/commit/8945fa85b03a99a83ea156efe448ac16d7c42844

**Comment by SarcrofS** [comment]
Is correcly lSDLmain? Because I have libSDL2main, every "lSDL..." is "libSDL" when downloaded from https://www.libsdl.org/download-2.0.php

Returned erros don't found lSDL when compiling.

**Comment by SarcrofS** [comment]
this files lSDL are .dll? .h? .a? I download from the provided link and not found with this names.

**Comment by SarcrofS** [comment]
LOL, now I learn how to "install" SDL in MinGW, copying files from "lib" to "lib" and keep making this to same name folders. See in this http://lazyfoo.net/SDL_tutorials/lesson01/windows/devcpp/index.php

But where put the binarie .dll downloaded from libsdl.org ?!

I try to compile and are not found others things
![cmder3](https://cloud.githubusercontent.com/assets/28415584/26126867/9fc32db2-3a5d-11e7-921d-836bb913db41.png)

**Comment by dorkster** [comment]
The SDL linked on that page is 1.2. You want 2.0. You'll need:

1. the Runtime binaries for your CPU (probably x64); these will be the DLLs you keep next to your flare.exe
2. the Development libraries for mingw; this includes the header files and ".a" libs

* SDL - https://www.libsdl.org/download-2.0.php
* SDL_image - https://www.libsdl.org/projects/SDL_image/
* SDL_mixer - https://www.libsdl.org/projects/SDL_mixer/
* SDL_ttf - https://www.libsdl.org/projects/SDL_ttf/

**Comment by SarcrofS** [comment]
```
c:/mingw/bin/../lib/gcc/mingw32/5.3.0/../../../../mingw32/bin/ld.exe: cannot find -lSDLmain
collect2.exe: error: ld returned 1 exit status
```

I install all this but have this error.

**Comment by dorkster** [comment]
I recall linking with SDLmain being weird on windows. You may also need to link with `-lmwindows`

**Comment by SarcrofS** [comment]
I tried:
`g++ -I C:\MinGW\include\SDL src\*.cpp -o flare.exe -lmingw32 -lSDLmain -lmwindows -lSDL2 -lSDL2_image -lSDL2_mixer -lSDL2_ttf`
and
`g++ -I C:\MinGW\include\SDL src\*.cpp -o flare.exe -lmingw32 -lmwindows -lSDL2 -lSDL2_image -lSDL2_mixer -lSDL2_ttf`
![cmdder](https://cloud.githubusercontent.com/assets/28415584/26128248/2e868b9e-3a62-11e7-909d-5780d6c257a3.png)

**Comment by dorkster** [comment]
Okay, forget about `-lmwindows`.

You *do* in fact need to use `-lSDL2main` instead of `-lSDLmain`. The file `C:\\MinGW\lib\libSDL2main.a` should exist if you copied everything over correctly.

**Comment by SarcrofS** [comment]
![error](https://cloud.githubusercontent.com/assets/28415584/26128917/789d2998-3a64-11e7-88ab-3d81d1a44f69.png)

A lot of undefined and one error.

**Comment by dorkster** [comment]
Make sure that all the other `libSDL*.a` files are in `C:\MinGW\lib`. They should match the `-l` commands in the command line.

**Comment by SarcrofS** [comment]
![libn](https://cloud.githubusercontent.com/assets/28415584/26129260/ea3b8bd4-3a65-11e7-8f61-721a58b08684.png)

**Comment by dorkster** [comment]
Just to be sure, did you also copy over the contents of the "bin", "include" and "share" directories?

**Comment by SarcrofS** [comment]
I copy from the downlods link of you provide me, ctrl+c ctrl+v bin files in MinGW/bin, libs in libs, include in include and share in share.

**Comment by dorkster** [comment]
You might need to put the SDL2 dll files in `C:\MinGW\bin`

**Comment by SarcrofS** [comment]
I put SDL2.dll, SDL2_image.dll SDL2_mixer.dll and SDL2_ttf.dll, any other need ?!

**Comment by dorkster** [comment]
No that's it. I'm not sure why linking isn't working for you. I can successfully compile with g++ in my Windows 7 VM and I can't tell how our MinGW/SDL2 installations are different.

**Comment by SarcrofS** [comment]
Thanks for try a lot with me @dorkster, you are a good person, keep this. 👍 

I will go out for now, maybe when back tomorrow I try more.

**Comment by igorko** [comment]
why don't you use cmake under windows/mingw?

**Comment by SarcrofS** [comment]
Because cmake have a lot of entries and I dont know how put then correcly, so return in error.
